Add initial test for multiplex worker corruption

restingbull opened the pull request.

restingbull committed:
Add initial test for multiplex worker corruption

restingbull committed:
fix starlark

restingbull committed:
Rename File to Paths for sanity

Bencodes commented on src/test/kotlin/io/bazel/kotlin/integration/MultiplexTest.kt:
```diff
@@ -0,0 +1,216 @@
+          }
+          1 -> {
+            one {
+              kotlin("One.kt") {
```
Just writing `One${round}.kt` and then referencing that in `Two.kt1 like `class Two : One${round}` should be enough to repro.

restingbull replied:
Testing says I need to make it bigger... :/

rockwotj commented on src/test/kotlin/io/bazel/kotlin/integration/WriteWorkspace.kt:
```diff
@@ -0,0 +1,364 @@
+package io.bazel.kotlin.integration
+
```
There is a lot of stuff here - at what point does it make sense to pull in something like [bazel-integration-testing](https://github.com/bazelbuild/bazel-integration-testing/blob/master/java/README.md) and then just have some lightweight kotlin extensions on those methods so you can write in the style of MultiplexTest...

restingbull replied:
Soon, I think. This was a first pass.

restingbull committed:
Add remove to opening a workspace.

restingbull committed:
Per rules_kotlin/issues/652

Switch to java implementation of the PersistentWorker, move the coroutine impl to side pending investigation

restingbull committed:
* Switch to Java specific multiplex handling
* Set IO to track thread local when capturing stdout and stderr
* Small threading fixes in the test structure (threads push things harder tha coroutines)

restingbull committed:
Fix build file to exclude coroutine again.

Bencodes commented:
@restingbull I ran some tests in our codebase with the worker java implementation and wasn't able to reproduce the issue. I was however able to reproduce it almost instantly switching back to the coroutines implementation.

rockwotj commented:
There are a bunch of thread local variables in the compiler using the IR backend: https://github.com/JetBrains/kotlin/blob/54cca882797de3d7861fcddae0214ad802937669/compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/SymbolTable.kt#L360-L374

I'm not sure if something like that could be the source of this error, but it seems like it if switching to a stable thread caller vs coroutine that may switch threads - thread local seems like a likely cause of these issues. Maybe if you use a coroutine dispatch that always switches threads at yield points may be a good way to force and see if that is the issue? (although I'm not sure how to do that off the top of my head)

rockwotj commented:
Looks like you could do that with a dispatcher that always returns true for `isDispatchNeeded` and then have one that for dispatch creates a new thread for every task...

Docs:
https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-dispatcher/is-dispatch-needed.html
https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-dispatcher/dispatch.html


Alternatively I wonder if you ran on a dispatcher with only a single thread if you wouldn't see corruption...

restingbull committed:
add jdk converter

restingbull committed:
revert
